databuilder/query_engines/sqlite.py

```diff
@@ -166,14 +166,19 @@ def get_sql_select_column(self, node):
         source = self.get_sql(node.source)
         return source.c[node.name]
 
-    # We have to apply caching here otherwise we generate distinct objects representing
-    # the same table and this confuses SQLAlchemy into generating queries with ambiguous
-    # table references
     @get_sql.register(SelectTable)
     @get_sql.register(SelectPatientTable)
-    @cache
     def get_sql_select_table(self, node):
-        return self.backend.get_table_expression(node.name)
+        return self.table_from_backend(node.name)
+
+    # We have to apply caching here otherwise we generate distinct objects representing
+    # the same table and this confuses SQLAlchemy into generating queries with ambiguous
+    # table references. We don't want to cache on the entire node because that holds a
+    # schema as well as the name, but the schema is irrelevant to the query engine and so
+    # should not be considered in the cache.
+    @cache
+    def table_from_backend(self, name):
+        return self.backend.get_table_expression(name)
```

Comment:
I'm happy to leave this in place for now, but I think we'll need an alternative fix further down the road. Although we don't currently use the schema during execution I think that will change. For instance, we may use it to define appropriate SQLAlchemy column types rather than requiring the backend to specify these as we currently do.

I'd argue that really these are mis-generated queries as instances of the same table should logically have the same schema. But at present we have nothing that enforces that and it's not obvious where the best place to do so would be.

Reply:
Ooh, yes. I meant to flag this as something that I wasn't really happy with. And this is definitely only viable while we're not using the schema in the engine.

I agree with your other comment below which makes the only driver for this go away. I'll create an issue to cover all this.

Reply:
https://github.com/opensafely-core/databuilder/issues/462